tools/clang-display-context/ClangDispCont.cpp

```diff
@@ -0,0 +1,169 @@
+using namespace clang;
+using namespace clang::tooling;
+
+static cl::OptionCategory ClangDispContCategory("clang-disp-context options");
+
+static cl::opt<uint> Line(
+    "line",
+    cl::desc("Line number of the issue's position. \n"),
+    cl::cat(ClangDispContCategory));
+
+static cl::opt<uint> Col(
+    "col",
+    cl::desc("Column number of the issue's position. \n"),
+    cl::cat(ClangDispContCategory));
+
+class DisplayContextConsumer : public ASTConsumer {
+public:
+  DisplayContextConsumer(ASTContext &Context) : Ctx(Context) {}
+
+  ~DisplayContextConsumer() {
+    // Flush results to standard output.
+    outs() << IssueString;
+  }
+
+  virtual void HandleTranslationUnit(ASTContext &Ctx) {
+    if (!Ctx.getDiagnostics().hasErrorOccurred())
+      handleDecl(Ctx.getTranslationUnitDecl());
+  }
+
+private:
+  template <typename T> 
+  bool checkIfInteresting(const T *N);
+  void handleDecl(const Decl *D);
+  void handleStmt(const Stmt *S);
+  void setEnclosingDecl(const Decl *D);
+  void setIssueString(const std::string &S);
+  
+  ASTContext &Ctx;
+  FullSourceLoc IssueLoc;
+  const Decl *EnclosingDecl;
+  std::string IssueString;
+};
+
+template <typename T>
+bool DisplayContextConsumer::checkIfInteresting(const T *Node) {
+  // Nodes are only interesting if they contain the line with the issue
+  if (!Node)
+    return false;
+
+  FullSourceLoc L1 = Ctx.getFullLoc(Node->getLocStart());
+  FullSourceLoc L2 = Ctx.getFullLoc(Node->getLocEnd());
+
+  return L1.isValid() && L2.isValid() &&
+         Line <= L2.getExpansionLineNumber() &&
+         L1.getExpansionLineNumber() <= Line;
+}
+
+void DisplayContextConsumer::handleDecl(const Decl *D) {
+  if (!D)
+    return;
+
+  if (const auto *TD = dyn_cast<TemplateDecl>(D))
+    D = TD->getTemplatedDecl();
+  
+  if (const auto *DC = dyn_cast<DeclContext>(D)) {
+    for (const Decl *D : DC->decls()) {
+      if (!checkIfInteresting<Decl>(D))
+        continue;
+      handleDecl(D);
+      return;
+    }
+  }
+  
+  setEnclosingDecl(D);
+  IssueLoc = Ctx.getFullLoc(D->getLocation()).getExpansionLoc();
+  
+  handleStmt(D->getBody());
+  setIssueString(clang::GetIssueString(Ctx.getSourceManager(), IssueLoc, "",
+                                       "", EnclosingDecl, Ctx.getLangOpts()));
+}
+
+void DisplayContextConsumer::handleStmt(const Stmt *S) {
+  if (!checkIfInteresting<Stmt>(S))
+    return;
+
+  IssueLoc = Ctx.getFullLoc(S->getLocStart()).getExpansionLoc();
+  
+  if (const LambdaExpr *LE = dyn_cast<LambdaExpr>(S))
+    EnclosingDecl = LE->getCallOperator();
+
+  for (const Stmt *Child : S->children())
+    handleStmt(Child);
+}
+
+void DisplayContextConsumer::setEnclosingDecl(const Decl *D) {
+  // If the issue is located in a DeclaratorDecl which is not
+  // a FunctionDecl, traverse up until the enclosing context is found.
+  // This function aims to solve issues especially with VarDecls.
+  while (dyn_cast<DeclaratorDecl>(D) && !dyn_cast<FunctionDecl>(D)) {
```

Comment:
Would this handle properly e.g. the `TagDecl`s? They can be `DeclContext` as well. What is the case also with transparent contexts? 
```
  /// isTransparentContext - Determines whether this context is a
  /// "transparent" context, meaning that the members declared in this
  /// context are semantically declared in the nearest enclosing
  /// non-transparent (opaque) context but are lexically declared in
  /// this context. For example, consider the enumerators of an
  /// enumeration type:
  /// @code
  /// enum E {
  ///   Val1
  /// };
  /// @endcode
  /// Here, E is a transparent context, so its enumerator (Val1) will
  /// appear (semantically) that it is in the same context of E.
  /// Examples of transparent contexts include: enumerations (except for
  /// C++0x scoped enums), and C++ linkage specifications.

```